### PR TITLE
Feat user signup

### PR DESCRIPTION
**Pull request by Ofektom**
### Description
Develop and integrate a seamless user sign up feature to enhance user engagement and onboarding experience. This feature should include user registration, account creation, and authentication processes.
### Acceptance Criteria
- [ ]  Email and password, firstname and lastname fields should be required
- [ ] Password validation should include a minimum of 8 alphanumeric characters
- [ ] Password should be hashed before being stored in the database
- [ ] A 422 error response code is returned on failed validations
- [ ] 201 status code is returned after successful signup
- [ ] JWT token should be generated for the user upon successful sign up and returned as part of the response payload
### Expected Outcome
- The user input data should be successfully sent to the backend, or the user will receive an error message if unsuccessful
### Requirements
**Corresponding database fileds for:**
- First Name
- Last Name
- Email Address
- Password
```
Table users {
  id integer [primary key]
  identifier uuid
  first_name varchar
  last_name varchar
  email varchar
  password varchar
  created_at timestamp
  updated_at timestamp
}
```
**Hash user password**
- Develop validation logic for user-submitted data.
- Validate username/email format (e.g., email address must be a valid format).
- Ensure password meets complexity requirements (e.g., minimum length, character types).
- Implement confirmation password field validation to ensure passwords match.
- Display clear error messages for invalid inputs.
### **User Registration [POST] `/api/v1/auth/register`**
1. Registration Endpoint
- Given a request with valid user details (ie `firstName`, `lastName`, `email`, `password`), when the user registers, then the system should create a new user account with a 201 created status code.
2. Unique Email
- Given a email that already exists, when the user tries to register, then the system should return a 400 bad request error status with an appropriate error message
3. Password Encryption
- Given a user registration request, when the user registers, then the system should store the password in an encrypted form.
**Request**
```
POST /api/v1/auth/register
{
  "firstName": String,
  "lastName": String,
  "email": String,
  "password": String,
}
```
**Successful Response**
```
{
 "statusCode": 201
  "message": String,
  "data" : {
        "token": string
        "user": {
          "id": String,
          "first_name": string
          "last_name": string
          "email": String,
          "created_at": Date
        },
   }
}
```
**Error Response**
```
{
  "message": String,
  "error": String,
  "status_code": Int,
}
```

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/config/DatabaseSeeder.java [hunk 1]
```diff
@@ -1,124 +1,124 @@
-package hng_java_boilerplate.config;
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/config/DatabaseSeeder.java:1]: Why did you comment out the database seeder?

File: src/main/java/hng_java_boilerplate/config/WebSecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+
+    @Bean
+    public AuthenticationManager authenticationManager(AuthenticationConfiguration authenticationConfiguration) throws Exception {
+        return authenticationConfiguration.getAuthenticationManager();
+    }
+
+
+    @Bean//bcryptPasswordEncoder is enabled for spring security hashing/salting of user's password information
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/config/WebSecurityConfig.java:60]: Remove All inline comments

File: src/main/java/hng_java_boilerplate/user/controller/UserController.java [hunk 3]
```diff
@@ -12,13 +10,12 @@
...
 
 import javax.crypto.BadPaddingException;
-import java.util.Optional;
 
 @RestController
 @RequiredArgsConstructor
 @RequestMapping("/api/v1/users")
 public class UserController {
-    private final UserService userService;
+    private final UserServiceImpl userService;
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/user/controller/UserController.java:18]: Why did you rename the `UserService`?

File: src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    @NotBlank(message = "Last name is required")
+    private String lastName;
+
+    @Email(message = "Email should be valid")
+    @NotBlank(message = "Email is required")
+    private String email;
+
+    @NotBlank(message = "Password is required")
+    @ValidPassword
+//    @Size(min = 8, message = "Password must be at least 8 characters long and contain alphanumeric characters")
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java:27]: Remove this comment

File: src/main/java/hng_java_boilerplate/user/dto/response/ErrorResponse.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ErrorResponse {
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/user/dto/response/ErrorResponse.java:11]: Remove extra space

File: src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+
+
+@Component
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java:23]: Remove extra all extra spaces in this file

File: src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    @Autowired
+    public JwtAuthenticationFilter(JwtUtils utils, @Lazy UserServiceImpl userService) {
+        this.utils = utils;
+        this.userService = userService;
+    }
+
+
+    @Override
+    protected void doFilterInternal(
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java:37]: format the code in this method properly

File: src/main/java/hng_java_boilerplate/util/JwtUtils.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+@Component
+public class JwtUtils {
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/util/JwtUtils.java:26]: Remove all extra spaces in this file

File: src/main/java/hng_java_boilerplate/util/PasswordValidator.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package hng_java_boilerplate.util;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/util/PasswordValidator.java:2]: Remove this extra space

File: src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
+package hng_java_boilerplate.user.end2end_test;
+
```
Comment (NdubuisiJr) [inline src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java:2]: Remove this extra space

File: src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    @Autowired
+    private UserRepository userRepository;
+
+    @Autowired
+    private WebApplicationContext context;
+
+    @BeforeEach
+    void setUp() {
+        mockMvc = MockMvcBuilders.webAppContextSetup(context).build();
+        userRepository.deleteAll(); // Clear the database before each test
```
Comment (NdubuisiJr) [inline src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java:40]: Which db are you clearing before each test?

File: src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+
+    @Test
+    void registerUser_InvalidPassword() throws Exception {
+        SignupDto signupDto = new SignupDto();
+        signupDto.setFirstName("John");
+        signupDto.setLastName("Doe");
+        signupDto.setEmail("john.doe@example.com");
+        signupDto.setPassword("pass"); // Invalid password
```
Comment (NdubuisiJr) [inline src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java:91]: Remove in-line comment

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

File: src/main/java/hng_java_boilerplate/config/DatabaseSeeder.java (hunk 1 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/config/DatabaseSeeder.java:1]: because my update to the user class would affect it since I added more field and that would require updating the database seeder class. Do I update the database seeder class to accomodate my changes?

File: src/main/java/hng_java_boilerplate/user/controller/UserController.java (hunk 3 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/user/controller/UserController.java:18]: He didn't have interface class so when I changed the UserService class to an interface, I transferred the initial codes in the UserService class into UserServiceImpl class which implements the UserService interface

File: src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java (hunk 4 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java:27]: Okay will do

File: src/main/java/hng_java_boilerplate/config/WebSecurityConfig.java (hunk 2 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/config/WebSecurityConfig.java:60]: Okay, will do

File: src/main/java/hng_java_boilerplate/user/dto/response/ErrorResponse.java (hunk 5 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/user/dto/response/ErrorResponse.java:11]: WIll do

File: src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java (hunk 6 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java:23]: Okay

File: src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java (hunk 7 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/util/JwtAuthenticationFilter.java:37]: Will do

File: src/main/java/hng_java_boilerplate/util/JwtUtils.java (hunk 8 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/util/JwtUtils.java:26]: Noted

File: src/main/java/hng_java_boilerplate/util/PasswordValidator.java (hunk 9 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/util/PasswordValidator.java:2]: Okay

File: src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java (hunk 10 above)
Comment (Ofektom) [reply src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java:2]: Okay

File: src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java (hunk 11 above)
Comment (Ofektom) [reply src/test/java/hng_java_boilerplate/user/end2end_test/AuthControllerTest.java]: I have removed that

**Review by Ofektom — COMMENTED** [review]
Issues resolved

**Review by Ofektom — COMMENTED** [review]
Done

**Comment by Ofektom** [comment]
@Am0du Please help me review my pull request

File: src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+import jakarta.validation.constraints.Size;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class SignupDto {
+    @NotBlank(message = "Full name is required")
```
Comment (Olaoluwa22) [inline src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java:15]: In line 15, you mean to say First name is required.

**Review by billmal071 — APPROVED** [review]
LGTM

File: src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java (hunk 13 above)
Comment (Ofektom) [reply src/main/java/hng_java_boilerplate/user/dto/request/SignupDto.java]: @Olaoluwa22 Done

**Comment by Ofektom** [comment]
@Olaoluwa22 please check my work again

**Review by NdubuisiJr — APPROVED** [review]

**Review by Olaoluwa22 — APPROVED** [review]
